rllib/algorithms/ppo/utils/env_runner.py

```diff
@@ -0,0 +1,471 @@
+                    initial_observation=obs[i],
+                    # TODO (simon): Check, if this works for the default
+                    # stateful encoders.
+                    initial_state={k: s[i] for k, s in states.items()},
+                )
+        # Do not reset envs, but instead continue in already started episodes.
+        else:
+            # Pick up stored observations and srtates from previous timesteps.
+            obs = np.stack([eps.observations[-1] for eps in self._episodes])
+            # COmpile the initial state for each batch row (vector sub_env):
+            # If episode just started, use the model's initial state, in the
+            # other case use the state stored last in the Episode.
+            states = {
+                k: np.stack(
+                    [
+                        initial_states[k][i] if eps.states is None else eps.states[k]
+                        for i, eps in enumerate(self._episodes)
+                    ]
+                )
+                for k in initial_states.keys()
+            }
+
+        # Loop through env in enumerate.(self._episodes):
+        ts = 0
+        # pbar = tqdm(
+        #     total=num_timesteps, desc=f"Sampling {num_timesteps} timesteps ..."
+        # )
+        while ts < num_timesteps:
+            # Act randomly.
+            if random_actions:
+                actions = self.env.action_space.sample()
+            # Compute an action using the RLModule.
+            else:
+                # Note, RLModule `forward()` methods expect `NestedDict`s.
+                batch = {
+                    STATE_IN: tree.map_structure(
+                        lambda s: tf.convert_to_tensor(s),
+                        states,
+                    ),
+                    SampleBatch.OBS: tf.convert_to_tensor(obs),
+                }
+                # Explore or not.
+                if explore:
+                    # TODO (simon) Implement this for MARL.
+                    fwd_out = self.module[DEFAULT_POLICY_ID].forward_exploration(batch)
+                    # `self.module` is a MARL module.
+                    action_dist_cls = foreach_module(
+                        self.module,
+                        lambda m, mid: (mid, m.get_exploration_action_dist_cls()),
+                    )
+                else:
+                    fwd_out = self.module[DEFAULT_POLICY_ID].forward_inference(batch)
+                    # `self.module` is a MARL module.
+                    action_dist_cls = foreach_module(
+                        self.module,
+                        lambda m, mid: (mid, m.get_inference_action_dist_cls()),
+                    )
+
+                action_dist_cls = {mid: dist_cls for mid, dist_cls in action_dist_cls}
+                action_dist = action_dist_cls[DEFAULT_POLICY_ID].from_logits(
+                    fwd_out[SampleBatch.ACTION_DIST_INPUTS]
+                )
+
+                actions = action_dist.sample().numpy()
```

Comment:
Actually, can we move the entire exploration logic now into the PPO modules?
We should not bother with specific exploration behavior here in the EnvRunner.

The API is currently as follows.
See: https://docs.ray.io/en/latest/rllib/rllib-rlmodule.html#writing-custom-single-agent-rl-modules (scroll a little down for a description of the return values for `_forward_inference/exploration`)